### PR TITLE
Cluster, AzureCluster and MachinePool conditions

### PR DESCRIPTION
**Pull request by nprokopic**
## What's in the box

`Cluster` conditions in a `clusterconditions` handler:

- `ProviderInfrastructureReady`: True when `AzureCluster` is Ready
- `NodePoolsReady`: True when all node pools (all `MachinePool` CRs) are ready
- `Ready`: True when `ProviderInfrastructureReady && NodePoolsReady`
- `Creating`: set to `True` when the cluster is initially created, set to `False` when the creation is completed.
- `Upgrading`: set to `True` when the upgrade has been started, set to `False` when the upgrade has been completed.

`AzureCluster` conditions in `azureclusterconditions` handler:

- `ResourceGroupReady`: `True` when Azure resource group is created.
- `VPNGatewayReady`: `True` when Azure VPN Gateway deployment has been completed successfully. Here we will add a check of actual VPN gateway resource and its state (as successful deployment is not a guarantee that the VPN Gateway is fully working).
- `Ready`: `True` when `ResourceGroupReady && VPNGatewayReady`

`MachinePool` conditions in `machinepoolconditions` handler:

- `ProviderInfrastructureReady`: `True` when `AzureMachinePool` is ready.
- `Ready`: `True` when `ProviderInfrastructureReady` is true. Here we will add checking of `Node` CR `Ready` condition for all node pool nodes.

`AzureMachinePool` conditions (handler will be added, currently set in existing handlers):

- `DeploymentSucceeded`: `True` when node pool deployment has been successful.
- `Ready`: `True` when `DeploymentSucceeded`. This is being updated and it will depend on `VMSSReady` and `SubnetReady` conditions.

### During cluster creation

`Cluster` CR:

```
$ k get cluster -n org-giantswarm m54jd -o yaml | yq r - "status.conditions"     
- lastTransitionTime: "2020-10-27T09:06:08Z"
  message: VPN Gateway deployment vpn-template is not found, check back in few minutes
  reason: DeploymentNotFound @ /m54jd
  severity: Warning
  status: "False"
  type: Ready
- lastTransitionTime: "2020-10-27T09:05:58Z"
  status: "True"
  type: Creating
- lastTransitionTime: "2020-10-27T09:13:29Z"
  message: 0 of 1 completed
  reason: Deploying @ /5cxc4
  severity: Warning
  status: "False"
  type: NodePoolsReady
- lastTransitionTime: "2020-10-27T09:06:08Z"
  message: VPN Gateway deployment vpn-template is not found, check back in few minutes
  reason: DeploymentNotFound @ /m54jd
  severity: Warning
  status: "False"
  type: ProviderInfrastructureReady
- lastTransitionTime: "2020-10-27T09:05:58Z"
  message: '{"message":"Upgrade not started","release_version":"13.0.0-nikolaconditions"}'
  reason: UpgradeNotStarted
  severity: Info
  status: "False"
  type: Upgrading
```

`AzureCluster` CR:
```
$ k get azurecluster -n org-giantswarm m54jd -o yaml | yq r - "status.conditions"
- lastTransitionTime: "2020-10-27T09:06:05Z"
  message: VPN Gateway deployment vpn-template is not found, check back in few minutes
  reason: DeploymentNotFound @ /m54jd
  severity: Warning
  status: "False"
  type: Ready
- lastTransitionTime: "2020-10-27T09:06:15Z"
  status: "True"
  type: ResourceGroupReady
- lastTransitionTime: "2020-10-27T09:06:05Z"
  message: VPN Gateway deployment vpn-template is not found, check back in few minutes
  reason: DeploymentNotFound
  severity: Warning
  status: "False"
  type: VPNGatewayReady
```

`MachinePool` CR:

```
$ k get machinepool -n org-giantswarm 5cxc4 -o yaml | yq r - "status.conditions"
- lastTransitionTime: "2020-10-27T09:11:36Z"
  message: Node pool deployment is in progress.
  reason: Deploying @ /5cxc4
  severity: Warning
  status: "False"
  type: Ready
- lastTransitionTime: "2020-10-27T09:11:36Z"
  message: Node pool deployment is in progress.
  reason: Deploying
  severity: Warning
  status: "False"
  type: ProviderInfrastructureReady
```

`AzureMachinePool` CR:

```
$ k get azuremachinepool -n org-giantswarm 5cxc4 -o yaml | yq r - "status.conditions"
- lastTransitionTime: "2020-10-27T09:09:44Z"
  message: Node pool deployment is in progress.
  reason: Deploying
  severity: Warning
  status: "False"
  type: Ready
- lastTransitionTime: "2020-10-27T09:09:48Z"
  message: Current deployment provisioning status is Running.
  reason: ProvisioningStateRunning
  severity: Warning
  status: "False"
  type: DeploymentSucceeded
```

### After cluster creation has been completed

`Cluster` CR:

```
$ k get cluster -n org-giantswarm m54jd -o yaml | yq r - "status.conditions"     
- lastTransitionTime: "2020-10-27T09:39:50Z"
  status: "True"
  type: Ready
- lastTransitionTime: "2020-10-27T09:39:50Z"
  message: Cluster creation has been completed in 33m53.914359348s
  reason: CreationCompleted
  severity: Info
  status: "False"
  type: Creating
- lastTransitionTime: "2020-10-27T09:18:44Z"
  status: "True"
  type: NodePoolsReady
- lastTransitionTime: "2020-10-27T09:39:50Z"
  status: "True"
  type: ProviderInfrastructureReady
- lastTransitionTime: "2020-10-27T09:05:58Z"
  message: '{"message":"Upgrade not started","release_version":"13.0.0-nikolaconditions"}'
  reason: UpgradeNotStarted
  severity: Info
  status: "False"
  type: Upgrading
```

`AzureCluster` CR:

```
$ k get azurecluster -n org-giantswarm m54jd -o yaml | yq r - "status.conditions"
- lastTransitionTime: "2020-10-27T09:39:12Z"
  status: "True"
  type: Ready
- lastTransitionTime: "2020-10-27T09:06:15Z"
  status: "True"
  type: ResourceGroupReady
- lastTransitionTime: "2020-10-27T09:39:12Z"
  status: "True"
  type: VPNGatewayReady
```

`MachinePool` CR:

```
$ k get machinepool -n org-giantswarm 5cxc4 -o yaml | yq r - "status.conditions"
- lastTransitionTime: "2020-10-27T09:16:21Z"
  status: "True"
  type: Ready
- lastTransitionTime: "2020-10-27T09:16:21Z"
  status: "True"
  type: ProviderInfrastructureReady
```

`AzureMachinePool` CR:

```
$ k get azuremachinepool -n org-giantswarm 5cxc4 -o yaml | yq r - "status.conditions"
- lastTransitionTime: "2020-10-27T09:14:38Z"
  status: "True"
  type: Ready
- lastTransitionTime: "2020-10-27T09:14:38Z"
  status: "True"
  type: DeploymentSucceeded
```

### CODE REVIEW

**Review by whites11 — APPROVED** [review]
LGTM

File: service/controller/resource/azureclusterconditions/conditionready.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	corev1 "k8s.io/api/core/v1"
+	capz "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+	capi "sigs.k8s.io/cluster-api/api/v1alpha3"
+	capiconditions "sigs.k8s.io/cluster-api/util/conditions"
+)
+
+func (r *Resource) ensureReadyCondition(ctx context.Context, azureCluster *capz.AzureCluster) error {
+	r.logDebug(ctx, "setting condition Ready")
+
+	// Note: This is an incomplete implementation that checks only resource
```
Comment (whites11) [inline service/controller/resource/azureclusterconditions/conditionready.go:17]: Is the final implementation still TODO?

Comment (nprokopic) [reply service/controller/resource/azureclusterconditions/conditionready.go:17]: yes

Comment (nprokopic) [reply service/controller/resource/azureclusterconditions/conditionready.go:17]: But this is good enough as is for what we needed for node pools release.

Comment (nprokopic) [reply service/controller/resource/azureclusterconditions/conditionready.go:17]: I will continue covering more resources with conditions, so it's more precise, but we should not have any issues for Creating and Upgrading that were needed for the front end to show the correct cluster state.

**Pull request merged**